trantor/net/TcpConnection.h

```diff
@@ -58,20 +56,49 @@ class TcpConnection
     virtual void shutdown() = 0;
     virtual void forceClose() = 0;
     virtual EventLoop *getLoop() = 0;
+    template <typename T>
+    void setContext(T &&context)
```

Comment:
On a more programmatic and less build-system note, I question the moving into this function then forward constructing it style.  That's mandating an API that should not necessarily be necessary and will fail with many user types.  I would instead opt for full inline construction with perfect forwarding if you want to construct it inline, or better yet, just take a `std::shared_ptr` and make the user themselves construct it however they want (this wouldn't even need to be a template then).

Reply:
The API can accept copyable or movable type, if user want to store some un-copyable and un-movable objects in the context(such as std::mutex), he should swap it with smart pointer  and then use `setContext`.
Compared with a `setContext(const std::shared_ptr<void>&)` interface, this one has less annoying type conversions in use (that is done in the getContext function).
As far as type compatibility is concerned, the current interface can accept copyable left-value references or movable right-value references, bare pointers, smart pointers or string literals. I think it can be adapted to most scenarios.